### PR TITLE
[GEARPUMP-252] return application status to client

### PR DESCRIPTION
**Pull request by huafengw**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the commit message is formatted like:
   `[GEARPUMP-<Jira issue #>] Meaningful description of pull request` 
 - [x] Make sure tests pass via `sbt clean test`.
 - [x] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

**Comment by manuzhang** [comment]
@huafengw I don't see much benefits with these changes. Do you have further refactoring in plan ?

File: core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala [hunk 1]
```diff
@@ -139,4 +138,47 @@ case class ExecutorContext(
...
+    executorAkkaConfig: Config = ConfigFactory.empty())
+
+// Not using Enumeration here because upickle does not support it
+sealed trait ApplicationStatus {
+  def isTerminalStatus: Boolean
+}
+
+object ApplicationStatus {
+  case object Pending extends ApplicationStatus {
+    override def toString: String = "pending"
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: You may define this only once in the superclass and pass in the status string from subclasses.

Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: FYI. http://underscore.io/blog/posts/2014/09/03/enumerations.html

File: core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala [hunk 2]
```diff
@@ -139,4 +138,47 @@ case class ExecutorContext(
 case class ExecutorJVMConfig(
     classPath: Array[String], jvmArguments: Array[String], mainClass: String,
     arguments: Array[String], jar: Option[AppJar], username: String,
-    executorAkkaConfig: Config = ConfigFactory.empty())
\ No newline at end of file
+    executorAkkaConfig: Config = ConfigFactory.empty())
+
+// Not using Enumeration here because upickle does not support it
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: There are also some other reasons to prefer case object. Please check the reference below.

**Comment by manuzhang** [comment]
@huafengw seems the changes not only relate to a simple removal. Could you rephrase your PR title ?

File: core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationMetaData.scala [hunk 3]
```diff
@@ -19,18 +19,20 @@
...
 /**
- * This state for single application, it is be distributed across the masters.
+ * The meta data of an application, which stores the crucial infomation of how to launch
+ * the application, like the application jar file location. This data is distributed
+ * across the masters.
  */
-case class ApplicationState(
-    appId: Int, appName: String, attemptId: Int, app: AppDescription, jar: Option[AppJar],
-    username: String, state: Any) extends Serializable {
+case class ApplicationMetaData(appId: Int, attemptId: Int, app: AppDescription,
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationMetaData.scala]: better name `app` to `appDesc` or `appDescription`

File: core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationMetaData.scala [hunk 4]
```diff
@@ -19,18 +19,20 @@
...
+    jar: Option[AppJar], username: String) extends Serializable {
 
   override def equals(other: Any): Boolean = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationMetaData.scala]: why is `override` required ?

File: core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationMetaData.scala (hunk 3 above)
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationMetaData.scala]: why do we need to override `equals` and `hashCode` methods of this class ?

File: core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala [hunk 5]
```diff
@@ -19,20 +19,34 @@
...
-    config: Config = null)
-  extends AppMasterRegisterData
+    config: Config = ConfigFactory.empty(),
+    status: ApplicationStatus = ApplicationStatus.Nonexist) {
+
+  def onAppMasterRegister(appMaster: ActorRef, worker: ActorRef): ApplicationRuntimeInfo = {
+    this.copy(appMaster = appMaster, worker = worker)
+  }
+
+  def onActived(timeStamp: TimeStamp): ApplicationRuntimeInfo = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala]: I think the verb is `Activate`

File: core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala [hunk 6]
```diff
@@ -19,20 +19,34 @@
...
+    user: String = "",
     submissionTime: TimeStamp = 0,
     startTime: TimeStamp = 0,
     finishTime: TimeStamp = 0,
-    config: Config = null)
-  extends AppMasterRegisterData
+    config: Config = ConfigFactory.empty(),
+    status: ApplicationStatus = ApplicationStatus.Nonexist) {
+
+  def onAppMasterRegister(appMaster: ActorRef, worker: ActorRef): ApplicationRuntimeInfo = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala]: Can we align the naming here ? e.g.

* onAppMasterRegistered
* onAppMasterActivated
* onAppMasterTerminated

File: core/src/main/scala/org/apache/gearpump/cluster/master/AppManager.scala [hunk 7]
```diff
@@ -305,50 +312,41 @@ private[cluster] class AppManager(kvService: ActorRef, launcher: AppMasterLaunch
...
-
-      appMasterRegistry += appId -> {
-        val (ref, info) = appMasterRegistry(appId)
-        (ref, info.copy(finishTime = System.currentTimeMillis()))
-      }
-      kvService ! PutKV(MASTER_GROUP, MASTER_STATE,
-        MasterState(this.nextAppId, appMasterRegistry, activeAppMasters, deadAppMasters))
-      kvService ! DeleteKVGroup(appId.toString)
-    }
+  private def killAppMasterExecutor(appId: Int, worker: ActorRef): Unit = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/master/AppManager.scala]: `AppMasterExecutor` is confusing. I think `killAppMaster` will do

File: core/src/main/scala/org/apache/gearpump/cluster/master/AppManager.scala [hunk 8]
```diff
@@ -217,40 +206,56 @@ private[cluster] class AppManager(kvService: ActorRef, launcher: AppMasterLaunch
...
+              }
+            case failed@ApplicationStatus.Failed =>
+              killAppMasterExecutor(appId, appRuntimeInfo.worker)
+              updatedStatus = appRuntimeInfo.onTerminalStatus(timeStamp, failed)
+              appResultListeners.getOrElse(appId, List.empty).foreach{ client =>
+                client ! ApplicationFailed(appId, error)
+              }
+            case terminated@ApplicationStatus.Terminated =>
+              updatedStatus = appRuntimeInfo.onTerminalStatus(timeStamp, terminated)
+            case _ =>
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/master/AppManager.scala]: what is for ? Any comments ?

Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/master/AppManager.scala]: what is for ? Any comments ?

File: core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala [hunk 9]
```diff
@@ -18,11 +18,11 @@
 
 package org.apache.gearpump.cluster
 
-import scala.reflect.ClassTag
+import java.io._
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: please avoid star imports. You may configure the Intellij "auto start imports" count to a large number (e.g. 100).

File: core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala [hunk 10]
```diff
@@ -142,4 +138,25 @@ case class ExecutorContext(
 case class ExecutorJVMConfig(
     classPath: Array[String], jvmArguments: Array[String], mainClass: String,
     arguments: Array[String], jar: Option[AppJar], username: String,
-    executorAkkaConfig: Config = ConfigFactory.empty())
\ No newline at end of file
+    executorAkkaConfig: Config = ConfigFactory.empty())
+
+sealed abstract class ApplicationStatus(val status: String, val isTerminal: Boolean)
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: `Terminal` is usually the shell we use in Unix/Linux. Do you mean `Terminated` ?

Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: It's just mean applications' final status. `Terminated` means being shut down by user, `Finished` means the application finished by itself and `Failed` just means failed, these are all terminal status.

File: core/src/main/scala/org/apache/gearpump/cluster/ClusterMessage.scala [hunk 11]
```diff
@@ -155,9 +158,13 @@ object MasterToClient {
 
   /** Return the last error of this streaming application job */
   case class LastFailure(time: TimeStamp, error: String)
-}
 
-trait AppMasterRegisterData
+  sealed trait ApplicationResult
+
+  case class ApplicationFinished(appId: Int) extends ApplicationResult
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/ClusterMessage.scala]: What does `Finished` mean ? Do you want to say `Succeeded` ?

File: core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    appMaster: ActorRef = ActorRef.noSender,
+    worker: ActorRef = ActorRef.noSender,
+    user: String = "",
+    submissionTime: TimeStamp = 0,
+    startTime: TimeStamp = 0,
+    finishTime: TimeStamp = 0,
+    config: Config = ConfigFactory.empty(),
+    status: ApplicationStatus = ApplicationStatus.NOEXIST) {
+
+  def onAppMasterRegister(appMaster: ActorRef, worker: ActorRef): ApplicationRuntimeInfo = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala]: As commented previously, please align the naming style here, i.e. use the passive form of a verb.

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 13]
```diff
@@ -18,17 +18,21 @@
...
 
-import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
 import scala.util.{Failure, Success}
 import scala.concurrent.ExecutionContext.Implicits.global
 
-class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
+class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout,
+    system: ActorSystem) {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: Can we avoid passing `system` around ?

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 14]
```diff
@@ -40,6 +44,12 @@ class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
     }
   }
 
+  def waitUnilFinish(): ApplicationResult = {
+    val delegator = system.actorOf(Props(new AppResultListener(appId, master)))
+    Await.result(delegator.ask(WaitUntilFinish)(INF_TIMEOUT).asInstanceOf
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: What are the effects of the two Timeout values here ?

File: core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala (hunk 10 above)
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: We can have a `TerminalStatus` extending `ApplicationStatus` for pattern matching rather than checking the flag every time.

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 15]
```diff
@@ -40,6 +44,12 @@ class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
     }
   }
 
+  def waitUnilFinish(): ApplicationResult = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: I don't think `ApplicationResult` should be exposed to client since it's an internal interface.

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 16]
```diff
@@ -50,3 +60,23 @@ class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
   }
 }
 
+object RunningApplication {
+  // This magic number is derived from Akka's configuration, which is the maximum delay
+  private val INF_TIMEOUT = new Timeout(2147482 seconds)
+
+  private case object WaitUntilFinish
+
+  private class AppResultListener(appId: Int, master: ActorRef) extends Actor {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: I think you reimplemented the ask pattern.

File: core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala (hunk 10 above)
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: good idea

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala (hunk 13 above)
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: Unless creating another ActorSystem here.

File: core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala [hunk 17]
```diff
@@ -142,4 +139,26 @@ case class ExecutorContext(
...
+
+  case object ACTIVE extends ApplicationStatus("active")
+
+  case object SUCCEEDED extends ApplicationTerminalStatus("succeeded")
+
+  case object FAILED extends ApplicationTerminalStatus("failed")
+
+  case object TERMINATED extends ApplicationTerminalStatus("terminated")
+
+  case object NOEXIST extends ApplicationStatus("nonexist")
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/AppDescription.scala]: `noexist` or `nonexist` ?

File: core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+  def onAppMasterRegistered(appMaster: ActorRef, worker: ActorRef): ApplicationRuntimeInfo = {
+    this.copy(appMaster = appMaster, worker = worker)
+  }
+
+  def onAppMasterActivated(timeStamp: TimeStamp): ApplicationRuntimeInfo = {
+    this.copy(startTime = timeStamp, status = ApplicationStatus.ACTIVE)
+  }
+
+  def onTerminalStatus(timeStamp: TimeStamp, finalStatus: ApplicationTerminalStatus):
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/appmaster/ApplicationRuntimeInfo.scala]: `terminal` or `final` ?

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 19]
```diff
@@ -40,6 +44,16 @@ class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
     }
   }
 
+  def waitUnilFinish(): Unit = {
+    val delegator = system.actorOf(Props(new AppResultListener(appId, master)))
+    val result = ActorUtil.askActor[ApplicationResult](delegator, WaitUntilFinish, INF_TIMEOUT)
+    result match {
+      case failed: ApplicationFailed =>
+        throw failed.error
+      case _ =>
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: Better add some warning log for unmatched cases

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 20]
```diff
@@ -40,6 +44,16 @@ class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
     }
   }
 
+  def waitUnilFinish(): Unit = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: typo. `Unil` -> `Until`

Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: please add comments for this public methods.

File: core/src/main/scala/org/apache/gearpump/cluster/ClusterMessage.scala [hunk 21]
```diff
@@ -155,26 +158,20 @@ object MasterToClient {
 
   /** Return the last error of this streaming application job */
   case class LastFailure(time: TimeStamp, error: String)
-}
 
-trait AppMasterRegisterData
+  sealed trait ApplicationResult
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/ClusterMessage.scala]: how is `ApplicationResult` mapped to `ApplicationStatus` ? If both of them are internal interfaces, can we just return `ApplicationStatus` to the client ?

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 22]
```diff
@@ -40,6 +44,16 @@ class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
...
+    val delegator = system.actorOf(Props(new AppResultListener(appId, master)))
+    val result = ActorUtil.askActor[ApplicationResult](delegator, WaitUntilFinish, INF_TIMEOUT)
+    result match {
+      case failed: ApplicationFailed =>
+        throw failed.error
+      case _ =>
+    }
+  }
+
   def askAppMaster[T](msg: Any): Future[T] = {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: We already have one in `ActorUtil`. Why not reuse it ?

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 23]
```diff
@@ -18,17 +18,21 @@
...
 import org.apache.gearpump.util.ActorUtil
+import org.apache.gearpump.cluster.client.RunningApplication._
 
+import scala.concurrent.duration._
 import scala.concurrent.Future
 import scala.util.{Failure, Success}
 import scala.concurrent.ExecutionContext.Implicits.global
 
-class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
+class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout,
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: I think we need a user interface here. `RunningApplication` could be the default implementation. Plus, please add guide on how to use the interface.

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala (hunk 22 above)
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: The ActorUtil's askAppMaster will first `ResolveAppId` which is unnecessary here.

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala (hunk 23 above)
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: We can do it in following pr.

Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: sure

**Comment by manuzhang** [comment]
LGTM. Please fix unattended comments in following PRs.
